src/mi/Gaddress_validator.c

```diff
@@ -0,0 +1,349 @@
+static sig_atomic_t   _unw_address_validator_initialized = 0;
+static int _mem_validate_pipe[2] = {-1, -1};
+static bool (*_mem_validate_func) (unw_word_t, size_t);
+
+#pragma weak pthread_once
+
+
+#ifdef HAVE_PIPE2
+static void
+_do_pipe2 (int pipefd[2])
+{
+  int result UNUSED = pipe2 (pipefd, O_CLOEXEC | O_NONBLOCK);
+}
+#else
+static void
+_set_pipe_flags (int fd)
+{
+  int fd_flags = fcntl (fd, F_GETFD, 0);
+  int status_flags = fcntl (fd, F_GETFL, 0);
+
+  fd_flags |= FD_CLOEXEC;
+  fcntl (fd, F_SETFD, fd_flags);
+
+  status_flags |= O_NONBLOCK;
+  fcntl (fd, F_SETFL, status_flags);
+}
+
+static void
+_do_pipe2 (int pipefd[2])
+{
+  pipe (pipefd);
+  _set_pipe_flags(pipefd[0]);
+  _set_pipe_flags(pipefd[1]);
+}
+#endif
+
+
+static void
+_open_pipe (void)
+{
+  if (_mem_validate_pipe[0] != -1)
+    close (_mem_validate_pipe[0]);
+  if (_mem_validate_pipe[1] != -1)
+    close (_mem_validate_pipe[1]);
+
+  _do_pipe2 (_mem_validate_pipe);
+}
+
+
+/**
+ * Test is a memory address is valid by trying to write from it
+ * @param[in]  addr The address to validate
+ *
+ * @returns true of the memory address is valid (readable), false otherwise.
+ *
+ * This check works by using the address as a (one-byte) buffer in a
+ * write-to-pipe operation.  The write will fail if the memory is not in the
+ * process's address space and marked as readable.
+ */
+static bool
+_write_validate (unw_word_t addr)
+{
+  int ret = -1;
+  ssize_t bytes = 0;
+
+  do
+    {
+      char buf;
+      bytes = read (_mem_validate_pipe[0], &buf, 1);
+    }
+  while ( errno == EINTR );
+
+  if (!(bytes > 0 || errno == EAGAIN || errno == EWOULDBLOCK))
+    {
+      // re-open closed pipe
+      _open_pipe ();
+    }
+
+  do
+    {
+#ifdef HAVE_SYS_SYSCALL_H
+      /* use syscall insteadof write() so that ASAN does not complain */
+      ret = syscall (SYS_write, _mem_validate_pipe[1], addr, 1);
+#else
+      ret = write (_mem_validate_pipe[1], (void *)addr, 1);
+#endif
+    }
+  while ( errno == EINTR );
+
+  return ret > 0;
+}
+
+
+static bool
+_msync_validate (unw_word_t addr, size_t len)
+{
+  if (msync ( (void *)unw_page_start (addr), len, MS_ASYNC) != 0)
+    {
+      return false;
+    }
+
+  return _write_validate (addr);
+}
+
+
+#ifdef HAVE_MINCORE
+static bool
+_mincore_validate (unw_word_t addr, size_t len)
+{
+  unsigned char mvec[2]; /* Unaligned access may cross page boundary */
+
+  /* mincore could fail with EAGAIN but we conservatively return false
+     instead of looping. */
+  if (mincore ((void *)unw_page_start (addr), len, mvec) != 0)
+    {
+      return false;
+    }
+
+  return _write_validate (addr);
+}
+#endif
+
+
+static void
+_unw_address_validator_init(void)
+{
+  _open_pipe ();
+
+  /* Work out dynamically what memory validation function to use. */
+#ifdef HAVE_MINCORE
+  unsigned char present = 1;
+  size_t len = unw_page_size;
+  unw_word_t addr = unw_page_start((unw_word_t)&present);
+  unsigned char mvec[1];
```

Comment:
Could you please
#ifdef __FreeBSD__ 
  char mvec[1];
#else
  unsigned char mvec[1];
#endif

Reply:
Evidently I force-pushed to the wrong branch at some point since this file is not a part of this patchset.

Force-pushed the correct branch after rebasing on master.